nf_core/modules/modules_differ.py

```diff
@@ -0,0 +1,257 @@
+import difflib
+import enum
+import json
+import logging
+import os
+import re
+import shutil
+import tempfile
+from pathlib import Path
+
+from rich.console import Console
+from rich.syntax import Syntax
+
+import nf_core.utils
+from nf_core.utils import plural_s
+
+log = logging.getLogger(__name__)
+
+
+class ModulesDiffer:
+    """
+    Static class that provides functionality for computing diffs between
+    different instances of a module
+    """
+
+    class DiffEnum(enum.Enum):
+        """
+        Enumeration for keeping track of
+        the diff status of a pair of files
+        """
+
+        UNCHANGED = enum.auto()
+        CHANGED = enum.auto()
+        CREATED = enum.auto()
+        REMOVED = enum.auto()
+
+    @staticmethod
+    def get_module_diffs(from_dir, to_dir, for_git=True, dsp_from_dir=None, dsp_to_dir=None):
+        """
+        Compute the diff between the current module version
+        and the new version.
+
+        Args:
+            from_dir (strOrPath): The folder containing the old module files
+            to_dir (strOrPath): The folder containing the new module files
+            path_in_diff (strOrPath): The directory displayed containing the module
+                                      file in the diff. Added so that temporary dirs
+                                      are not shown
+            for_git (bool): indicates whether the diff file is to be
+                            compatible with `git apply`. If true it
+                            adds a/ and b/ prefixes to the file paths
+            dsp_from_dir (str | Path): The from directory to display in the diff
+            dsp_to_dir (str | Path): The to directory to display in the diff
+
+        Returns:
+            dict[str, (ModulesDiffer.DiffEnum, str)]: A dictionary containing
+            the diff type and the diff string (empty if no diff)
+        """
+        if dsp_from_dir is None:
+            dsp_from_dir = from_dir
+        if dsp_to_dir is None:
+            dsp_to_dir = to_dir
+        if for_git:
+            dsp_from_dir = Path("a", dsp_from_dir)
+            dsp_to_dir = Path("b", dsp_to_dir)
+
+        diffs = {}
+        # Get all unique filenames in the two folders.
+        # `dict.fromkeys()` is used instead of `set()` to preserve order
+        files = dict.fromkeys(os.listdir(to_dir))
+        files.update(dict.fromkeys(os.listdir(from_dir)))
+        files = list(files)
```

Comment:
Related to my comments on lines 97 and 110, although this is a punctual case, maybe we should check all folders recursively in case modules contain folders.